### PR TITLE
Add configurable word and character limits to questions

### LINKED ISSUE #268: Max word count
Add a configurable max word count for text input. This is needed for HackGT 6 launch.

### PR DESCRIPTION
**Pull request by petschekr**
* Admins can now add `maxWordCount`, `minWordCount`, `maxCharacterCount`, and `minCharacterCount` to fields that accept text
* Character counts use HTML5 form validation for better user experience (and won't allow sending of form until it passes)
* Word counts are represented live to the user as they type
* Word and character counts are checked server side to prevent sneaky users / crappy browsers from allowing bad inputs
* Word and character counts are stackable (can configure multiple limits for the same question)

Closes #268

### CODE REVIEW

File: client/js/application.ts [hunk 1]
```diff
@@ -126,3 +126,21 @@ for (let i = 0; i < inputsWithOther.length; i++) {
...
+
+let wordCountInputs = document.querySelectorAll("[data-max-word-count], [data-min-word-count]") as NodeListOf<HTMLInputElement | HTMLTextAreaElement>;
+for (let i = 0; i < wordCountInputs.length; i++) {
+	wordCountInputs[i].addEventListener("input", e => {
+		let target = e.target as HTMLInputElement | HTMLTextAreaElement;
+		let correspondingLabel = document.querySelector(`label[for="${target.id}"] > .current-count`);
+		if (!correspondingLabel) return;
+
+		let { maxWordCount, minWordCount } = target.dataset;
+		if (!maxWordCount && !minWordCount) return;
```
Comment (evan10s) [inline client/js/application.ts]:
```suggestion
		if (!maxWordCount && !minWordCount) { return; }
```

File: client/js/application.ts [hunk 2]
```diff
@@ -126,3 +126,21 @@ for (let i = 0; i < inputsWithOther.length; i++) {
 		}
 	});
 }
+
+let wordCountInputs = document.querySelectorAll("[data-max-word-count], [data-min-word-count]") as NodeListOf<HTMLInputElement | HTMLTextAreaElement>;
+for (let i = 0; i < wordCountInputs.length; i++) {
+	wordCountInputs[i].addEventListener("input", e => {
+		let target = e.target as HTMLInputElement | HTMLTextAreaElement;
```
Comment (evan10s) [inline client/js/application.ts]: seems like this can be `const`?

File: client/js/application.ts [hunk 3]
```diff
@@ -126,3 +126,21 @@ for (let i = 0; i < inputsWithOther.length; i++) {
...
+		let correspondingLabel = document.querySelector(`label[for="${target.id}"] > .current-count`);
+		if (!correspondingLabel) return;
```
Comment (evan10s) [inline client/js/application.ts]:
```suggestion
		if (!correspondingLabel) { return; }
```

File: client/js/application.ts [hunk 4]
```diff
@@ -126,3 +126,21 @@ for (let i = 0; i < inputsWithOther.length; i++) {
 		}
 	});
 }
+
+let wordCountInputs = document.querySelectorAll("[data-max-word-count], [data-min-word-count]") as NodeListOf<HTMLInputElement | HTMLTextAreaElement>;
+for (let i = 0; i < wordCountInputs.length; i++) {
+	wordCountInputs[i].addEventListener("input", e => {
+		let target = e.target as HTMLInputElement | HTMLTextAreaElement;
+		let correspondingLabel = document.querySelector(`label[for="${target.id}"] > .current-count`);
```
Comment (evan10s) [inline client/js/application.ts]: same here with `const`

File: client/js/application.ts [hunk 5]
```diff
@@ -126,3 +126,21 @@ for (let i = 0; i < inputsWithOther.length; i++) {
...
+		if (!correspondingLabel) return;
+
+		let { maxWordCount, minWordCount } = target.dataset;
```
Comment (evan10s) [inline client/js/application.ts]: and here? (`const`)

File: client/partials/form.html [hunk 6]
```diff
@@ -33,9 +33,24 @@
...
+				<small>(At least {{this.minCharacterCount}} characters)</small>
+			{{/if}}
+			{{#if this.maxCharacterCount}}
+				<small>(Max {{this.maxCharacterCount}} characters)</small>
+			{{/if}}
+			<small class="current-count"></small>
+		</label>
 		{{#ifCond this.type "textarea"}}
-			<textarea name="{{this.name}}" id="form-{{this.name}}" placeholder="{{this.placeholder}}" {{required this.required}}>{{this.value}}</textarea>
+			<textarea name="{{this.name}}" id="form-{{this.name}}" placeholder="{{this.placeholder}}" {{required this.required}} minlength="{{this.minCharacterCount}}" maxlength="{{this.maxCharacterCount}}" data-max-word-count="{{this.maxWordCount}}" data-min-word-count="{{this.minWordCount}}">{{this.value}}</textarea>
```
Comment (evan10s) [inline client/partials/form.html]: Does this still add the `minlength` and other attributes even if those aren't defined in the JSON file?

File: client/partials/form.html [hunk 7]
```diff
@@ -45,7 +60,7 @@
 					<input type="file" name="{{this.name}}" id="form-{{this.name}}" placeholder="{{this.placeholder}}" {{required this.required}} />
 				{{/if}}
 			{{else}}
-				<input type="{{this.type}}" name="{{this.name}}" id="form-{{this.name}}" placeholder="{{this.placeholder}}" {{required this.required}} value="{{this.value}}" />
+				<input type="{{this.type}}" name="{{this.name}}" id="form-{{this.name}}" placeholder="{{this.placeholder}}" {{required this.required}} minlength="{{this.minCharacterCount}}" maxlength="{{this.maxCharacterCount}}" data-max-word-count="{{this.maxWordCount}}" data-min-word-count="{{this.minWordCount}}" value="{{this.value}}" />
```
Comment (evan10s) [inline client/partials/form.html]: same here re: attrs

File: server/routes/api/user.ts [hunk 8]
```diff
@@ -132,31 +155,45 @@ function postApplicationBranchHandler(anonymous: boolean): (request: express.Req
...
+				return reportError(`Your response to "${question.label}" must have at least ${question.minCharacterCount} characters`);
+			}
+			if (question.maxCharacterCount && getQuestion("").length > question.maxCharacterCount) {
+				return reportError(`Your response to "${question.label}" cannot exceed ${question.maxCharacterCount} characters`);
+			}
+			let wordCount = getQuestion("").trim().split(/\s+/).length;
+			if (getQuestion("").trim().length === 0) {
+				wordCount = 0;
+			}
+			let wordCountPlural = wordCount === 1 ? "" : "s";
```
Comment (evan10s) [inline server/routes/api/user.ts]: use `const`

File: client/js/application.ts (hunk 2 above)
Comment (petschekr) [reply client/js/application.ts]: Good call. I'm too used to Rust at the moment where everything is immutable by default and it also uses `let`

File: client/partials/form.html (hunk 6 above)
Comment (petschekr) [reply client/partials/form.html]: Yeah it does
Is it worth adding a Handlebars partial to only add the attribute if it is set?

File: server/routes/api/user.ts [hunk 9]
```diff
@@ -132,31 +155,45 @@ function postApplicationBranchHandler(anonymous: boolean): (request: express.Req
...
 			}
-			if ((question.type === "select" || question.type === "radio") && Array.isArray(request.body[question.name]) && question.hasOther) {
+			if (question.minCharacterCount && getQuestion("").length < question.minCharacterCount) {
+				return reportError(`Your response to "${question.label}" must have at least ${question.minCharacterCount} characters`);
+			}
+			if (question.maxCharacterCount && getQuestion("").length > question.maxCharacterCount) {
+				return reportError(`Your response to "${question.label}" cannot exceed ${question.maxCharacterCount} characters`);
+			}
+			let wordCount = getQuestion("").trim().split(/\s+/).length;
+			if (getQuestion("").trim().length === 0) {
```
Comment (ehsanmasdar) [inline server/routes/api/user.ts]: getQuestion("") is not truthy @petschekr. This returns undefined for file upload fields and which crashes registration

**Pull request merged**